ZEngine/src/Helper.cpp

```diff
@@ -0,0 +1,879 @@
+        VkPipelineLayout&                                           pipeline_layout,
+        VkRenderPass&                                               render_pass,
+        const Pipelines::GraphicRendererPipelineStateSpecification& specification)
+    {
+        VkPipeline out_graphic_pipeline{VK_NULL_HANDLE};
+
+        VkGraphicsPipelineCreateInfo graphic_pipeline_create_info = {};
+        graphic_pipeline_create_info.sType                        = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO;
+
+        std::vector<VkPipelineShaderStageCreateInfo> shader_create_info_collection;
+        std::vector<VkShaderModule>                  shader_module_collection;
+
+        for (const auto& shader_information : specification.ShaderInfo)
+        {
+            shader_create_info_collection.push_back(shader_information.ShaderStageCreateInfo);
+            shader_module_collection.push_back(shader_information.ShaderModule);
+        }
+
+        graphic_pipeline_create_info.stageCount = shader_create_info_collection.size();
+        graphic_pipeline_create_info.pStages    = shader_create_info_collection.data();
+
+        graphic_pipeline_create_info.pVertexInputState   = &(specification.VertexInputStateCreateInfo);
+        graphic_pipeline_create_info.pInputAssemblyState = &(specification.InputAssemblyStateCreateInfo);
+        graphic_pipeline_create_info.pViewportState      = &(specification.ViewportStateCreateInfo);
+        graphic_pipeline_create_info.pRasterizationState = &(specification.RasterizationStateCreateInfo);
+        graphic_pipeline_create_info.pMultisampleState   = &(specification.MultisampleStateCreateInfo);
+        graphic_pipeline_create_info.pDepthStencilState  = &(specification.DepthStencilStateCreateInfo);
+        graphic_pipeline_create_info.pColorBlendState    = &(specification.ColorBlendStateCreateInfo);
+        graphic_pipeline_create_info.pDynamicState       = &(specification.DynamicStateCreateInfo);
+        graphic_pipeline_create_info.layout              = pipeline_layout;
+        graphic_pipeline_create_info.renderPass          = render_pass;
+        graphic_pipeline_create_info.subpass             = 0;
+        graphic_pipeline_create_info.basePipelineHandle  = VK_NULL_HANDLE; // Optional
+        graphic_pipeline_create_info.basePipelineIndex   = -1;             // Optional
+        graphic_pipeline_create_info.flags               = 0;              // Optional
+        graphic_pipeline_create_info.pNext               = nullptr;        // Optional
+
+        ZENGINE_VALIDATE_ASSERT(
+            vkCreateGraphicsPipelines(device.GetNativeDeviceHandle(), VK_NULL_HANDLE, 1, &graphic_pipeline_create_info, nullptr, &out_graphic_pipeline) == VK_SUCCESS,
+            "Failed to create Graphics Pipeline")
+
+        // Cleanup ShaderModules
+        for (auto& shader_module : shader_module_collection)
+        {
+            vkDestroyShaderModule(device.GetNativeDeviceHandle(), shader_module, nullptr);
```

Comment:
Not Vulkan specific but I would say it is a bit of an API contract problem if you pass in the pipline specification object as const& but then call destroy on the shader module handles contained in that specification. I also don't think you really need the ShaderModule member at all in the specification since the create infos already contain the handles.

Reply:
Fixed, thanks a lot !